### PR TITLE
Invoice rebase

### PR DESCRIPTION
**Pull request by Ibrahimhizeoui**

### CODE REVIEW

**Comment by Nyholm** [comment]
Thank you. Can you fix the failing tests?

File: src/Api/Customer.php [hunk 1]
```diff
@@ -92,25 +87,19 @@ public function create(Model $customer)
...
      * @return mixed|\Psr\Http\Message\ResponseInterface
      *
+     * @throws NotFoundException
      * @throws ValidationException
      */
     public function update(int $customerNo, Model $costumer)
     {
         $response = $this->httpPut('/customer/'.$customerNo, $costumer->toArray());
-
+        $body = $response->getBody()->__toString();
```
Comment (Nyholm) [inline src/Api/Customer.php]: This is debug code. It should be removed

File: src/Api/Invoice.php [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+     *
+     * @return string|array
+     *
+     * @see https://billogram.com/api/documentation#billogram_call_create
+     */
+    public function search(array $param = [])
+    {
+        $param = array_merge(['page' => 1, 'page_size' => 100], $param);
+        $response = $this->httpget('/billogram', $param);
+        $body = $response->getBody()->__toString();
```
Comment (Nyholm) [inline src/Api/Invoice.php]: This is debug code. It should be removed

File: src/Api/Item.php [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+namespace Billogram\Api;
+
+use Billogram\Exception\Domain\ValidationException;
+use Billogram\Model\Item\Item as Model;
+use Billogram\Model\Item\Items;
+
+/**
+ * @author Ibrahim Hizeoui <ibrahimhizeoui@gmail.com>
+ */
+class Item extends HttpApi
```
Comment (Nyholm) [inline src/Api/Item.php]: This should not be here.. It is a different feature.

File: src/CachedResponseClient.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+            $url = str_replace($this->apiKey, '[apikey]', $url);
+        }
+        $file = sprintf('%s/%s_%s', $this->cacheDir, $host, sha1($url));
+        if (is_file($file) && is_readable($file)) {
+            $header = unserialize(file_get_contents("tests/.cache/headers.txt"));
+            return new Response(200, ['Content-Type' => $header['Content-Type'][0]], (new StreamFactory())->createStream(unserialize(file_get_contents($file))));
+        }
+        $response = $this->delegate->sendRequest($request);
+        file_put_contents($file, serialize($response->getBody()->getContents()));
+        file_put_contents("tests/.cache/headers.txt", serialize(($response->getHeaders())));
```
Comment (Nyholm) [inline src/CachedResponseClient.php]: Wouldn't this file be overwritten by the next response?

File: composer.json [hunk 5]
```diff
@@ -20,7 +20,10 @@
         "php-http/discovery": "^1.0",
         "php-http/httplug": "^1.0",
         "php-http/multipart-stream-builder": "^1.0",
-        "php-http/client-common": "^1.5"
+        "php-http/client-common": "^1.5",
+        "psr/http-message-implementation": "^1.0",
+        "php-http/mock-client": "^1.0",
+        "nyholm/psr7": "^0.2.2"
```
Comment (Nyholm) [inline composer.json]: Why are these in the required section?

File: src/BaseTestCase.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        $client->addResponse(new Response($statusCode, [], $body));
+        return $client;
+    }
+
+    /**
+     * Get a mocked HTTP client where you may do tests on the request.
+     * @param callable $requestCallback
+     * @return \PHPUnit_Framework_MockObject_MockObject
+     */
+    protected function getMockedHttpClientCallback(callable $requestCallback)
```
Comment (Nyholm) [inline src/BaseTestCase.php]: Is this method ever used?

File: src/CachedResponseClient.php (hunk 4 above)
Comment (Ibrahimhizeoui) [inline src/CachedResponseClient.php]: yes but I thought we have the same content-type ??

Comment (Nyholm) [inline src/CachedResponseClient.php]: If we are sure we always have the same content type, then we do not need the store it in a file in the first place.

But what about other headers?

File: src/BaseTestCase.php (hunk 6 above)
Comment (Ibrahimhizeoui) [inline src/BaseTestCase.php]: yes i will delete it

File: composer.json (hunk 5 above)
Comment (Ibrahimhizeoui) [inline composer.json]: i have copied from your repo I can put it in require dev

File: src/CachedResponseClient.php (hunk 4 above)
Comment (Ibrahimhizeoui) [inline src/CachedResponseClient.php]: yes we have the same content type and i think we don't need other headers

File: src/Api/Invoice.php (hunk 2 above)
Comment (Ibrahimhizeoui) [inline src/Api/Invoice.php]: ok can I change it to here ??
throw new DomainExceptions\NotFoundException($response->getBody()->__toString());

File: src/Api/Item.php (hunk 3 above)
Comment (Ibrahimhizeoui) [inline src/Api/Item.php]: yes I will update the item model

File: composer.json [hunk 7]
```diff
@@ -18,21 +18,27 @@
...
         "phpunit/phpunit": "^6.2",
         "php-http/message-factory": "^1.0",
         "php-http/curl-client": "^1.6",
         "php-http/message": "^1.0",
         "guzzlehttp/psr7": "^1.3",
-        "nyholm/nsa": "^1.0"
+        "nyholm/nsa": "^1.0",
+        "php-http/multipart-stream-builder": "^1.0",
+        "php-http/client-common": "^1.5",
+        "psr/http-message-implementation": "^1.0",
```
Comment (Nyholm) [inline composer.json]: Remove

File: composer.json [hunk 8]
```diff
@@ -18,21 +18,27 @@
...
     "require-dev": {
         "phpunit/phpunit": "^6.2",
         "php-http/message-factory": "^1.0",
         "php-http/curl-client": "^1.6",
         "php-http/message": "^1.0",
         "guzzlehttp/psr7": "^1.3",
-        "nyholm/nsa": "^1.0"
+        "nyholm/nsa": "^1.0",
+        "php-http/multipart-stream-builder": "^1.0",
+        "php-http/client-common": "^1.5",
```
Comment (Nyholm) [inline composer.json]: Move this and stream builder to require

File: headers [hunk 9]
```diff
@@ -0,0 +1 @@
...;}s:7:"Expires";a:1:{i:0;s:29:"Mon, 31 Jul 2017 14:48:20 GMT";}s:4:"Vary";a:1:{i:0;s:23:"Cookie, Accept-Encoding";}s:13:"Cache-Control";a:1:{i:0;s:9:"max-age=0";}s:10:"Set-Cookie";a:1:{i:0;s:108:"dtCookie=6$2C2A838D534366AE85C761B20DCA2867|RUM+Default+Application|1; Path=/; Domain=.billogram.com; secure";}s:23:"Content-Security-Policy";a:1:{i:0;s:59:"default-src https: data: wss: 'unsafe-inline' 'unsafe-eval'";}s:15:"X-Frame-Options";a:1:{i:0;s:10:"SAMEORIGIN";}s:16:"X-Xss-Protection";a:1:{i:0;s:13:"1; mode=block";}s:22:"X-Content-Type-Options";a:1:{i:0;s:7:"nosniff";}s:25:"Strict-Transport-Security";a:1:{i:0;s:44:"max-age=31536000; includeSubdomains; preload";}s:15:"Public-Key-Pins";a:1:{i:0;s:487:"pin-sha256="+qVQvnt5yb253ksCTDShLi4YF0vAPSNklg/lZZBVP/8="; pin-sha256="wdVM/85nxkG2Y1/J/qYfsIuCEsj8jzERJ3Aw7yAeDSw="; pin-sha256="KXTSUWRog1JWXRETuXpaNxwBb70GhK2DPbxT48pqGM0="; pin-sha256="cbRybe3tXRUTuqwtSimQpF9niYVqDxVCgzCBOyGLJgo="; pin-sha256="lPY3Z7kFEInh8+8Qgsinuj2eME1bQOVNnVc13s3TYI0="; pin-sha256="b8WqC0bsigc3akPYzfSUBKnI0qRfz3G4EWFwhS8305k="; pin-sha256="UPYaBL6pD2gw8CIQPLZh16/DG+16iIrph3ZUy6fKx7w="; pin-sha256="IOBCdzKsvdCTAb/L7+3VIMtG/doHybMnJL4HFShbzsM="; max-age=2592000";}}
```
Comment (Nyholm) [inline headers]: Remove this file

File: headers.txt [hunk 10]
```diff
@@ -0,0 +1 @@
...;}s:7:"Expires";a:1:{i:0;s:29:"Mon, 31 Jul 2017 14:51:18 GMT";}s:4:"Vary";a:1:{i:0;s:23:"Cookie, Accept-Encoding";}s:13:"Cache-Control";a:1:{i:0;s:9:"max-age=0";}s:10:"Set-Cookie";a:1:{i:0;s:108:"dtCookie=6$6F85BB7B8B446A87169281FBB3BD33AF|RUM+Default+Application|1; Path=/; Domain=.billogram.com; secure";}s:23:"Content-Security-Policy";a:1:{i:0;s:59:"default-src https: data: wss: 'unsafe-inline' 'unsafe-eval'";}s:15:"X-Frame-Options";a:1:{i:0;s:10:"SAMEORIGIN";}s:16:"X-Xss-Protection";a:1:{i:0;s:13:"1; mode=block";}s:22:"X-Content-Type-Options";a:1:{i:0;s:7:"nosniff";}s:25:"Strict-Transport-Security";a:1:{i:0;s:44:"max-age=31536000; includeSubdomains; preload";}s:15:"Public-Key-Pins";a:1:{i:0;s:487:"pin-sha256="+qVQvnt5yb253ksCTDShLi4YF0vAPSNklg/lZZBVP/8="; pin-sha256="wdVM/85nxkG2Y1/J/qYfsIuCEsj8jzERJ3Aw7yAeDSw="; pin-sha256="KXTSUWRog1JWXRETuXpaNxwBb70GhK2DPbxT48pqGM0="; pin-sha256="cbRybe3tXRUTuqwtSimQpF9niYVqDxVCgzCBOyGLJgo="; pin-sha256="lPY3Z7kFEInh8+8Qgsinuj2eME1bQOVNnVc13s3TYI0="; pin-sha256="b8WqC0bsigc3akPYzfSUBKnI0qRfz3G4EWFwhS8305k="; pin-sha256="UPYaBL6pD2gw8CIQPLZh16/DG+16iIrph3ZUy6fKx7w="; pin-sha256="IOBCdzKsvdCTAb/L7+3VIMtG/doHybMnJL4HFShbzsM="; max-age=2592000";}}
```
Comment (Nyholm) [inline headers.txt]: Remove this file

File: src/Api/Customer.php [hunk 11]
```diff
@@ -37,14 +39,13 @@ public function search(array $param = [])
      * @param int   $customerNo
      * @param array $param
      *
-     * @see https://billogram.com/api/documentation#customers_fetch
+     * @return mixed|\Psr\Http\Message\ResponseInterface
```
Comment (Nyholm) [inline src/Api/Customer.php]: Why was the link removed? 
Is this `@return` better then previous one?

File: src/Api/Invoice.php (hunk 2 above)
Comment (Nyholm) [inline src/Api/Invoice.php]: Yes, remove this.

**Comment by Nyholm** [comment]
Thank you

**Pull request merged**